File: benchmarks/Java/richards/RBObject.java
package som.richards;


public class RBObject {
  public Packet append(final Packet packet, final Packet queueHead) {
    packet.setLink(RBObject.noWork());
    if (RBObject.noWork() == queueHead) {
      return packet;
    }

    Packet mouse = queueHead;
    Packet link;
    while (RBObject.noWork() != (link = mouse.getLink())) {
      mouse = link;
    }
    mouse.setLink(packet);
    return queueHead;
  }

  private static int    DeviceA;
  private static int    DeviceB;
  private static int    DevicePacketKind;
  private static int    HandlerA;
  private static int    HandlerB;
  private static int    Idler;
  private static TaskControlBlock NoTask;
  private static Packet NoWork;
  private static int    Worker;
  private static int    WorkPacketKind;

  public static TaskControlBlock noTask() { return NoTask; }
  public static int    idler()  { return Idler;  }
  public static Packet noWork() { return NoWork; }
  public static int    worker() { return Worker; }
  public static int    workPacketKind() { return WorkPacketKind; }
  public static int    handlerA() { return HandlerA; }
  public static int    handlerB() { return HandlerB; }
  public static int    deviceA()  { return DeviceA;  }
  public static int    deviceB()  { return DeviceB;  }
  public static int    devicePacketKind() { return DevicePacketKind; }

  public static void initialize() {
    DeviceA          = 5;
    DeviceB          = 6;
    DevicePacketKind = 1;
    HandlerA         = 3;
    HandlerB         = 4;
    Idler            = 1;
    NoWork           = null;
    NoTask           = null;
    Worker           = 2;
    WorkPacketKind   = 2;
  }
}

File: benchmarks/Rust/src/richards/rb_object.rs
use crate::richards::packet::PacketBox;
use crate::richards::task_control_block::TaskControlBlock;

pub trait RBObject {
    fn append(mut packet: PacketBox, queue_head: Option<PacketBox>) -> PacketBox {
        packet.set_link(Self::NO_WORK);
        match queue_head {
            None => packet,
            Some(mut queue_head) => {
                queue_head.append_link(Some(packet));
                queue_head
            }
        }
    }

    const IDLER: usize = 0;
    const WORKER: usize = 1;
    const HANDLER_A: usize = 2;
    const HANDLER_B: usize = 3;
    const DEVICE_A: usize = 4;
    const DEVICE_B: usize = 5;
    const NUM_TYPES: usize = 6;

    const DEVICE_PACKET_KIND: usize = 0;
    const WORK_PACKET_KIND: usize = 1;

    const NO_WORK: Option<PacketBox> = None;
    const NO_TASK: Option<TaskControlBlock> = None;
}
